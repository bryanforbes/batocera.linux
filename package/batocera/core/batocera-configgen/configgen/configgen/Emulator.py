from __future__ import annotations

import collections.abc
import logging
import xml.etree.ElementTree as ET
from dataclasses import InitVar, dataclass, field
from typing import TYPE_CHECKING, Any, ClassVar, Literal, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_CONF, BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .settings.unixSettings import UnixSettings
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path
    from typing_extensions import deprecated

    from .gun import GunMapping

_logger = logging.getLogger(__name__)
_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: collections.abc.Mapping[str, Any]) -> None:
    """ Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if (key in destination and isinstance(destination[key], dict) and isinstance(value, collections.abc.Mapping)):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


@dataclass(slots=True)
class Emulator:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    name: str
    rom: InitVar[Path]

    config: dict[str, Any] = field(init=False)
    renderconfig: dict[str, Any] = field(init=False)

    def __post_init__(self, rom: Path, /) -> None:
        # read the configuration from the system name
        self.config = self.__get_system_config(
            DEFAULTS_DIR / "configgen-defaults.yml",
            DEFAULTS_DIR / "configgen-defaults-arch.yml"
        )
        if "emulator" not in self.config or not self.config["emulator"]:
            _logger.error("no emulator defined. exiting.")
            raise Exception("No emulator found")

        gsname = self.game_settings_name(rom)

        # load configuration from batocera.conf
        batoceraSettings = UnixSettings(BATOCERA_CONF)
        globalSettings = batoceraSettings.load_all('global')
        controllersSettings = batoceraSettings.load_all('controllers', includeName=True)
        systemSettings = batoceraSettings.load_all(self.name)
        folderSettings = batoceraSettings.load_all(f'{self.name}.folder["{rom.parent}"]')
        gameSettings = batoceraSettings.load_all(f'{self.name}["{gsname}"]')

        # add some other options
        displaySettings = batoceraSettings.load_all('display')
        for opt, value in displaySettings.items():
            self.config[f"display.{opt}"] = value

        # update config
        Emulator.update_configuration(self.config, controllersSettings)
        Emulator.update_configuration(self.config, globalSettings)
        Emulator.update_configuration(self.config, systemSettings)
        Emulator.update_configuration(self.config, folderSettings)
        Emulator.update_configuration(self.config, gameSettings)
        self.updateFromESSettings()
        _logger.debug("uimode: %s", self.config['uimode'])

        # forced emulators ?
        self.config["emulator-forced"] = False
        self.config["core-forced"] = False
        if "emulator" in globalSettings or "emulator" in systemSettings or "emulator" in gameSettings:
            self.config["emulator-forced"] = True
        if "core" in globalSettings or "core" in systemSettings or "core" in gameSettings:
            self.config["core-forced"] = True

        # update renderconfig
        self.renderconfig = {}
        if "shaderset" in self.config:
            if self.config["shaderset"] != "none":
                rendering_defaults = USER_SHADERS / "configs" / str(self.config["shaderset"]) / "rendering-defaults.yml"
                if rendering_defaults.exists():
                    self.renderconfig = self.__get_yaml_defaults(
                        rendering_defaults,
                        rendering_defaults.with_name("rendering-defaults-arch.yml")
                    )
                else:
                    rendering_defaults = BATOCERA_SHADERS / "configs" / str(self.config["shaderset"]) / "rendering-defaults.yml"
                    self.renderconfig = self.__get_yaml_defaults(
                        rendering_defaults,
                        rendering_defaults.with_name("rendering-defaults-arch.yml")
                    )
            elif self.config["shaderset"] == "none":
                rendering_defaults = BATOCERA_SHADERS / "configs" / "rendering-defaults.yml"
                self.renderconfig = self.__get_yaml_defaults(
                    rendering_defaults,
                    rendering_defaults.with_name("rendering-defaults-arch.yml")
                )

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        systemSettings = batoceraSettings.load_all(f'{self.name}-renderer')
        gameSettings = batoceraSettings.load_all(f'{self.name}["{gsname}"]-renderer')

        # es only allow to update systemSettings and gameSettings in fact for the moment
        Emulator.update_configuration(self.renderconfig, systemSettings)
        Emulator.update_configuration(self.renderconfig, gameSettings)

    @property
    def core(self) -> str:
        return self.config['core']

    @property
    def core_forced(self) -> bool:
        return self.config['core-forced']

    @property
    def emulator(self) -> str:
        return self.config['emulator']

    @property
    def emulator_forced(self) -> bool:
        return self.config['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.config['uimode']

    @property
    def show_fps(self) -> bool:
        return self.config['showFPS']

    def game_settings_name(self, rom: Path) -> str:
        # sanitize rule by EmulationStation
        # see FileData::getConfigurationName() on batocera-emulationstation
        gs_name = rom.name.replace('=','')
        gs_name = gs_name.replace('#','')
        _logger.info("game settings name: %s", gs_name)
        return gs_name

    def __get_yaml_defaults(self, default_yaml: Path, default_arch_yaml: Path, /) -> dict[str, Any]:
        with default_yaml.open() as f:
            defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

        arch_defaults: dict[str, Any] = {}
        if default_arch_yaml.exists():
            with default_arch_yaml.open() as f:
                loaded_arch_defaults = yaml.load(f, Loader=yaml.CLoader)

            if loaded_arch_defaults is not None:
                arch_defaults = loaded_arch_defaults

        config: dict[str, Any] = {}

        if 'default' in defaults:
            config = defaults['default']

        if 'default' in arch_defaults:
            _dict_merge(config, arch_defaults['default'])

        if self.name in defaults:
            _dict_merge(config, defaults[self.name])

        if self.name in arch_defaults:
            _dict_merge(config, arch_defaults[self.name])

        return config

    def __get_system_config(self, default_yaml: Path, default_arch_yaml: Path, /) -> dict[str, Any]:
        defaults = self.__get_yaml_defaults(default_yaml, default_arch_yaml)

        result: dict[str, Any] = { 'emulator': defaults['emulator'], 'core': defaults['core'] }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(result, defaults['options'])

        return result

    def has_option(self, key: str, /) -> bool:
        return key in self.config

    @overload
    def get_option(self, key: str, default: MissingType = ..., /) -> Any | MissingType:
        ...

    @overload
    def get_option(self, key: str, default: _T, /) -> Any | _T:
        ...

    def get_option(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.config.get(key, default)

    @overload
    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool:
        ...

    @overload
    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV:
        ...

    def get_option_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None) -> bool | _TV | _FV:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES
        else:
            return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_option_str(self, key: str, default: MissingType = ..., /) -> str | MissingType:
        ...

    @overload
    def get_option_str(self, key: str, default: str, /) -> str:
        ...

    def get_option_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_option_int(self, key: str, default: MissingType = ..., /) -> int | MissingType:
        ...

    @overload
    def get_option_int(self, key: str, default: int, /) -> int:
        ...

    def get_option_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_option_float(self, key: str, default: MissingType = ..., /) -> float | MissingType:
        ...

    @overload
    def get_option_float(self, key: str, default: float, /) -> float:
        ...

    def get_option_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.config.get(key, MISSING)

        if value is MISSING:
            return default

        return float(value)

    def option_items(self, /, *, starts_with: str | None = None) -> collections.abc.Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.config.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.config.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


    if TYPE_CHECKING:
        @deprecated('Use has_option()')
        def isOptSet(self, key: str) -> bool:
            ...

        @deprecated('Use get_option_bool()')
        def getOptBoolean(self, key: str) -> bool:
            ...

        @deprecated('Use get_option_str()')
        def getOptString(self, key: str) -> str:
            ...
    else:
        def isOptSet(self, key: str) -> bool:
            return key in self.config

        def getOptBoolean(self, key: str) -> bool:
            value = self.config.get(key)

            if isinstance(value, str):
                value = value.lower()

            return value in self.TRUE_VALUES

        def getOptString(self, key: str) -> str:
            if key in self.config:  # noqa: SIM102
                if self.config[key]:
                    return self.config[key]
            return ""

    def get_guns_borders_size(self, guns: GunMapping, /) -> str | None:
        borders_size = self.get_option_str("controllers.guns.borderssize", "medium")

        # overriden by specific options
        borders_mode = "normal"
        if (config_borders_mode := self.get_option_str("controllers.guns.bordersmode", "auto")) != "auto":
            borders_mode = config_borders_mode
        if (config_borders_mode := self.get_option_str("bordersmode", "auto")) != "auto":
            borders_mode = config_borders_mode

        # others are gameonly and normal
        if borders_mode == "hidden":
            return None

        if borders_mode == "force":
            return borders_size

        for gun in guns.values():
            if gun.need_borders:
                return borders_size

        return None

    def get_guns_borders_ratio(self, guns: GunMapping, /) -> str | None:
        # returns None to follow the bezel overlay size by default
        return self.get_option("controllers.guns.bordersratio", None)

    @staticmethod
    def update_configuration(config: dict[str, Any], settings: dict[str, Any], /) -> None:
        # ignore all values "default", "auto", "" to take the system value instead
        # ideally, such value must not be in the configuration file
        # but historically some users have them

        config.update({
            key: value for key, value in settings.items()
            if value and value != 'default' and value != 'auto'
        })

    # fps value is from es
    def updateFromESSettings(self):
        try:
            esConfig = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = esConfig.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib["value"] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            self.config['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = esConfig.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib["value"] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            self.config['uimode'] = uimode_value

        except Exception:
            self.config['showFPS'] = False
            self.config['uimode'] = "Full"
