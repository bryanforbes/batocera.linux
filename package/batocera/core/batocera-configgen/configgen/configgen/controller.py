from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
