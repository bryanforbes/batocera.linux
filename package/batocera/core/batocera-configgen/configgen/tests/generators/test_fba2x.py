from __future__ import annotations

from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.generators.fba2x.fba2xGenerator import Fba2xGenerator
from tests.generators.base import GeneratorBaseTest
from tests.mock_controllers import make_player_controller_list

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controller, Controllers
    from configgen.Emulator import Emulator


@pytest.mark.usefixtures('fs')
class TestFba2xGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[Fba2xGenerator]:
        return Fba2xGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'neogeo'

    @pytest.fixture
    def emulator(self) -> str:
        return 'fba2x'

    def test_generate(
        self,
        generator: Fba2xGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'neogeo' / 'rom.zip',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'fba' / 'fba2x.cfg').read_text() == snapshot(name='config')

    def test_generate_existing(
        self,
        generator: Fba2xGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(
            CONFIGS / 'fba' / 'fba2x.cfg',
            contents="""[Graphics]
DisplaySmoothStretch = 1

[Joystick]
SDLID_1 = 1

[Foo]
Bar = 2
""",
        )
        generator.generate(
            mock_system,
            ROMS / 'neogeo' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'fba' / 'fba2x.cfg').read_text() == snapshot

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'smooth': '0'},
            {'smooth': '1'},
            {'ratio': '4/3'},
            {'ratio': '16/9'},
            {'ratio': 'full'},
            {'shaders': 'foo'},
            {'shaders': 'scanlines'},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: Fba2xGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'neogeo' / 'rom.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'fba' / 'fba2x.cfg').read_text() == snapshot

    def test_generate_six_button_game(
        self,
        generator: Fba2xGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'neogeo' / 'sfa.zip',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'fba' / 'fba2x.cfg').read_text() == snapshot

    def test_generate_controllers(
        self,
        generator: Fba2xGenerator,
        mock_system: Emulator,
        generic_xbox_pad: Controller,
        gpio_controller_1: Controller,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'neogeo' / 'sfa.zip',
            make_player_controller_list(generic_xbox_pad, gpio_controller_1),
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'fba' / 'fba2x.cfg').read_text() == snapshot
