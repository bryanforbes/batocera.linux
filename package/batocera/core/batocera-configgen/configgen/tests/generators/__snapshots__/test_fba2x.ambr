# serializer version: 1
# name: TestFba2xGenerator.test_generate
  tuple(
    list([
      '/usr/bin/fba2x',
      '--configfile',
      '/userdata/system/configs/fba/fba2x.cfg',
      '--logfile',
      '/userdata/system/logs/fba2x.log',
      '/userdata/roms/neogeo/rom.zip',
    ]),
    dict({
    }),
  )
# ---
# name: TestFba2xGenerator.test_generate[config]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'ratio': '16/9'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 1
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'ratio': '4/3'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'ratio': 'full'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'shaders': 'foo'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'shaders': 'scanlines'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 1
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'smooth': '0'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_config[{'smooth': '1'}]
  '''
  [Graphics]
  DisplaySmoothStretch = 1
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_controllers
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = 1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  L_1 = 0
  Y_1 = 1
  X_1 = 2
  A_1 = 3
  B_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  L_2 = 1
  Y_2 = 0
  X_2 = 2
  A_2 = 3
  B_2 = 4
  R_2 = 5
  START_2 = 7
  SELECT_2 = 6
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_existing
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Foo]
  Bar = 2
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  Y_1 = 0
  X_1 = 1
  B_1 = 2
  A_1 = 3
  L_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_generate_six_button_game
  '''
  [Graphics]
  DisplaySmoothStretch = 0
  MaintainAspectRatio = 0
  DisplayEffect = 0
  RotateScreen = 0
  DisplayBorder = 0
  
  [Joystick]
  SDLID_1 = 0
  SDLID_2 = -1
  SDLID_3 = -1
  SDLID_4 = -1
  JA_UD_1 = 1
  JA_LR_1 = 0
  L_1 = 0
  Y_1 = 1
  X_1 = 2
  A_1 = 3
  B_1 = 4
  R_1 = 5
  START_1 = 9
  SELECT_1 = 8
  QUIT = 9
  HOTKEY = 8
  
  
  '''
# ---
# name: TestFba2xGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
    }),
    'name': 'fba2x',
  })
# ---
